sql/stats: implement canary stats with delayed deletion mechanism

ZhouXing19 opened the pull request:
Based on: #159192

**sql/stats: implement canary stats with delayed deletion mechanism**

This commit introduces a delayed deletion mechanism and dual-cache system
for table statistics to support canary stats functionality. Previously,
old statistics were immediately deleted when new ones were created, which
could cause query plan instability during statistics collection.

The implementation includes two main parts. First, a delayed deletion
mechanism replaces immediate deletion with a two-phase process:
`MarkDelayDelete` marks old statistics with `delayDelete=true` while
preserving them, and `DeleteStatsWithExpiredVisibility` removes
previously marked statistics only when the next canary cycle begins.

Second, the stats cache now maintains both `stats` and `stableStats`
collections along with `latestStatsTimestamp`. The `GetTableStats` API
determines which collection to return based on whether the canary window
has been exceeded since the latest stats were collected. This ensures
queries can use either the latest statistics (canary path) or preserve
existing statistics (stable path) while maintaining forecast
reconstructibility throughout the canary window.

For example, when creating statistic S6 for a column with S1-S5, S1 is
marked for deletion but preserved until S7 is created. The cache serves
S2-S6 for canary queries and S1-S5 for stable queries until the canary
window expires.

Release Notes (sql change): Table statistics now use a delayed deletion
mechanism to improve query plan stability during automatic statistics
collection. Old statistics are marked for delayed deletion rather than
immediately removed when new statistics are created.

ZhouXing19 committed:
sql: implement stats_as_of session variable for time-based statistics selection

This commit adds a new session variable `stats_as_of` that allows
controlling statistics selection based on a specific timestamp rather
than the current time. Previously, statistics selection was always
relative to the current wall clock time, making it difficult to get
consistent query plans for historical analysis or testing.

This feature is only for debugging and troubleshooting, and should not
be used in production.

The implementation is also integrated into the existing canary
statistics logic to respect the as-of timestamp when determining canary
window boundaries.

Release note (sql change): adds a new session variable `stats_as_of`
that allows controlling statistics selection based on a specific
timestamp rather than the current time.

cockroach-teamcity commented:
<details><summary><strong>🔴 Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 11.48m ±3% | 11.54m ±3% | ~ | p=0.461 n=15    |
| 🔴 **allocs/op** | 8.211k ±1% | 8.782k ±1% | +6.95% | p=0.000 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/93223aa/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/93223aa920a5b644bdbc2ed080fd128c9e69b3b9/bin/pkg_sql_tests benchdiff/93223aa/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/93223aa/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/6ddbffe/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/6ddbffe7fedb0f27a581335f12ffd6c96b4c8f7a/bin/pkg_sql_tests benchdiff/6ddbffe/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/6ddbffe/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=6ddbffe --new=93223aa ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.982m ±1% | 3.993m ±1% | ~ | p=0.870 n=15    |
| ⚪ **allocs/op** | 2.067k ±0% | 2.066k ±0% | ~ | p=0.593 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/93223aa/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/93223aa920a5b644bdbc2ed080fd128c9e69b3b9/bin/pkg_sql_tests benchdiff/93223aa/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/93223aa/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/6ddbffe/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/6ddbffe7fedb0f27a581335f12ffd6c96b4c8f7a/bin/pkg_sql_tests benchdiff/6ddbffe/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/6ddbffe/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=6ddbffe --new=93223aa ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 4.387m ±1% | 4.404m ±1% | ~ | p=0.174 n=15    |
| ⚪ **allocs/op** | 4.185k ±0% | 4.183k ±0% | ~ | p=0.370 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/93223aa/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/93223aa920a5b644bdbc2ed080fd128c9e69b3b9/bin/pkg_sql_tests benchdiff/93223aa/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/93223aa/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/6ddbffe/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/6ddbffe7fedb0f27a581335f12ffd6c96b4c8f7a/bin/pkg_sql_tests benchdiff/6ddbffe/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/6ddbffe/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=6ddbffe --new=93223aa ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/93223aa920a5b644bdbc2ed080fd128c9e69b3b9/20086225501-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/6ddbffe7fedb0f27a581335f12ffd6c96b4c8f7a/20086225501-1/\* old/
```

</details>

_built with commit: [93223aa920a5b644bdbc2ed080fd128c9e69b3b9](https://github.com/cockroachdb/cockroach/commit/93223aa920a5b644bdbc2ed080fd128c9e69b3b9)_

<!-- microbench-ci -->

ZhouXing19 committed:
sql: implement dual stats caches for table statistics for canary stats rollout

This change extends the table statistics cache to support two sets of
statistics: regular stats and stable stats. Previously, for each table,
the cache only maintained a single set of statistics. Now it maintains
both "freshest stats" and "stable stats", with the later starts from the
second latest full stats available from the system.table_statistics.

When stable=true and the age of the latest stats is within the canary
window, the cache returns stable stats instead of the freshest stats.

Release note: None

cockroach-teamcity commented:
<details><summary><strong>⚪ Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 11.81m ±4% | 11.66m ±3% | ~ | p=0.653 n=15    |
| ⚪ **allocs/op** | 8.213k ±0% | 8.208k ±1% | ~ | p=0.254 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/5dc68c1/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/5dc68c13e04783dbb6ba6c61b9a486affe1ef42e/bin/pkg_sql_tests benchdiff/5dc68c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/5dc68c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/6ddbffe/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/6ddbffe7fedb0f27a581335f12ffd6c96b4c8f7a/bin/pkg_sql_tests benchdiff/6ddbffe/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/6ddbffe/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=6ddbffe --new=5dc68c1 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.417m ±1% | 3.412m ±2% | ~ | p=1.000 n=15    |
| ⚪ **allocs/op** | 2.082k ±0% | 2.082k ±0% | ~ | p=0.961 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/5dc68c1/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/5dc68c13e04783dbb6ba6c61b9a486affe1ef42e/bin/pkg_sql_tests benchdiff/5dc68c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/5dc68c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/6ddbffe/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/6ddbffe7fedb0f27a581335f12ffd6c96b4c8f7a/bin/pkg_sql_tests benchdiff/6ddbffe/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/6ddbffe/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=6ddbffe --new=5dc68c1 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.774m ±2% | 3.752m ±0% | -0.57% | p=0.023 n=15    |
| ⚪ **allocs/op** | 4.179k ±0% | 4.181k ±0% | ~ | p=0.088 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/5dc68c1/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/5dc68c13e04783dbb6ba6c61b9a486affe1ef42e/bin/pkg_sql_tests benchdiff/5dc68c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/5dc68c1/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/6ddbffe/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/6ddbffe7fedb0f27a581335f12ffd6c96b4c8f7a/bin/pkg_sql_tests benchdiff/6ddbffe/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/6ddbffe/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=6ddbffe --new=5dc68c1 ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/5dc68c13e04783dbb6ba6c61b9a486affe1ef42e/20088161765-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/6ddbffe7fedb0f27a581335f12ffd6c96b4c8f7a/20088161765-1/\* old/
```

</details>

_built with commit: [5dc68c13e04783dbb6ba6c61b9a486affe1ef42e](https://github.com/cockroachdb/cockroach/commit/5dc68c13e04783dbb6ba6c61b9a486affe1ef42e)_

<!-- microbench-ci -->

ZhouXing19 committed:
sql/stats: introduce delayed deletion for table stats for canary stats tables

This commit introduces a delayed deletion mechanism to support canary
stats functionality. Previously, old statistics were immediately
deleted when new ones were created, which could cause query plan
instability during statistics collection.

The implementation includes two main parts. First, a delayed deletion
mechanism replaces immediate deletion with a two-phase process:
`DeleteExpiredStats` removes previously marked statistics, and
`MarkDelayDelete` marks stale statistics with `delayDelete=true` while
preserving them. These marked stats will be removed by
`DeleteExpiredStats` when another new stats is collected.

Note such 2-phase deletion only applies to tables with
`stats_canary_window > 0`, as in if they enabled the canary stats
rollout feature. For those with canary stats rollout disabled
(default), the immediate deletion behavior maintains the same.

For context, this is to ensures queries can use either the latest
statistics (canary path) or preserve existing statistics (stable path)
while maintaining forecast reconstructibility throughout the canary
window. (will be introduced in the next PR #159060)

Release note (sql change): For tables with canary stats mode enabled, we
now use a delayed deletion mechanism for its statistics to improve
query plan stability during automatic statistics collection. Old
statistics are marked for delayed deletion rather than immediately
removed when new statistics are created. For tables with canary stats
disabled (default), the stats retention logic is unchanged.

ZhouXing19 committed:
----REBASED FROM #159192----

ZhouXing19 committed:
sql/stats: delay delete for other columns

For stats that are not partial stats using extremes, the stats for columns
that are not directly involved in the new stats will be also removed.
We now implement the delayed deletion logic for these "other columns" as
well, to be aligned with what we do with the involved columns.

Release note: None

ZhouXing19 committed:
sql: introduce `ALTER TABLE ... PUSH STATISTICS [WITH EXPLICIT COLUMNS] <json>`

Informs: #150015

This change introduces a new ALTER TABLE PUSH STATISTICS syntax to
support deterministic testing of statistics collection evolution.
Currently, testing automatic statistics collection requires running
DMLs and polling system.table_statistics, which is non-deterministic
and cumbersome
(see pkg/sql/logictest/testdata/logic_test/distsql_automatic_stats).

While the existing ALTER TABLE INJECT STATISTICS completely replaces all
table statistics at once, it doesn't test the retention logic that
occurs during auto statistics collection. The new PUSH STATISTICS
syntax mimics `sampleAggregator.writeResults()` behavior by selectively
deleting old statistics for specific columns before inserting new ones,
preserving the natural evolution of the statistics table. This syntax
is exclusively for internal testing, particularly to support canary
stats rollout testing. It emphasizes the gradual evolution of
statistics and how the optimizer selects among multiple statistics as
new ones are continuously collected, rather than treating statistics
injection as a single point-in-time event.

Release note: None

ZhouXing19 committed:
accomodate push stats to delayed delete

ZhouXing19 committed:
----REBASED FROM #159845----

ZhouXing19 committed:
sql: add tests for canary stats rollout

Release note: None

cockroach-teamcity commented:
<details><summary><strong>⚪ Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 11.83m ±3% | 11.81m ±4% | ~ | p=0.250 n=15    |
| ⚪ **allocs/op** | 8.205k ±0% | 8.200k ±0% | ~ | p=0.976 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/e1208e8/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/e1208e80db17fbebdf231765558eb31284f17901/bin/pkg_sql_tests benchdiff/e1208e8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/e1208e8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/2b9d89f/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/2b9d89f2054b4269ae86516951407e84143928df/bin/pkg_sql_tests benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=2b9d89f --new=e1208e8 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.347m ±1% | 3.346m ±1% | ~ | p=0.653 n=15    |
| ⚪ **allocs/op** | 2.082k ±0% | 2.082k ±0% | ~ | p=0.985 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/e1208e8/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/e1208e80db17fbebdf231765558eb31284f17901/bin/pkg_sql_tests benchdiff/e1208e8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/e1208e8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/2b9d89f/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/2b9d89f2054b4269ae86516951407e84143928df/bin/pkg_sql_tests benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=2b9d89f --new=e1208e8 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.575m ±0% | 3.573m ±0% | ~ | p=0.902 n=15    |
| ⚪ **allocs/op** | 4.195k ±0% | 4.193k ±0% | ~ | p=0.132 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/e1208e8/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/e1208e80db17fbebdf231765558eb31284f17901/bin/pkg_sql_tests benchdiff/e1208e8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/e1208e8/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/2b9d89f/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/2b9d89f2054b4269ae86516951407e84143928df/bin/pkg_sql_tests benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=2b9d89f --new=e1208e8 ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/e1208e80db17fbebdf231765558eb31284f17901/20833322948-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/2b9d89f2054b4269ae86516951407e84143928df/20833322948-1/\* old/
```

</details>

_built with commit: [e1208e80db17fbebdf231765558eb31284f17901](https://github.com/cockroachdb/cockroach/commit/e1208e80db17fbebdf231765558eb31284f17901)_

<!-- microbench-ci -->

cockroach-teamcity commented:
<details><summary><strong>⚪ Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 11.50m ±4% | 11.53m ±4% | ~ | p=0.935 n=15    |
| ⚪ **allocs/op** | 8.200k ±2% | 8.196k ±2% | ~ | p=0.454 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/102a35d/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/102a35d68eb14b112199296a593162bc019000df/bin/pkg_sql_tests benchdiff/102a35d/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/102a35d/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/2b9d89f/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/2b9d89f2054b4269ae86516951407e84143928df/bin/pkg_sql_tests benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=2b9d89f --new=102a35d ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.435m ±1% | 3.408m ±1% | ~ | p=0.137 n=15    |
| ⚪ **allocs/op** | 2.082k ±0% | 2.083k ±0% | ~ | p=0.448 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/102a35d/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/102a35d68eb14b112199296a593162bc019000df/bin/pkg_sql_tests benchdiff/102a35d/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/102a35d/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/2b9d89f/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/2b9d89f2054b4269ae86516951407e84143928df/bin/pkg_sql_tests benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=2b9d89f --new=102a35d ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.793m ±3% | 3.808m ±1% | ~ | p=0.653 n=15    |
| ⚪ **allocs/op** | 4.180k ±0% | 4.180k ±0% | ~ | p=0.798 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/102a35d/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/102a35d68eb14b112199296a593162bc019000df/bin/pkg_sql_tests benchdiff/102a35d/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/102a35d/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/2b9d89f/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/2b9d89f2054b4269ae86516951407e84143928df/bin/pkg_sql_tests benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=2b9d89f --new=102a35d ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/102a35d68eb14b112199296a593162bc019000df/20834246928-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/2b9d89f2054b4269ae86516951407e84143928df/20834246928-1/\* old/
```

</details>

_built with commit: [102a35d68eb14b112199296a593162bc019000df](https://github.com/cockroachdb/cockroach/commit/102a35d68eb14b112199296a593162bc019000df)_

<!-- microbench-ci -->

cockroach-teamcity commented:
<details><summary><strong>⚪ Sysbench</strong> [SQL, 3node, oltp_read_write]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 11.04m ±3% | 11.04m ±3% | ~ | p=0.567 n=15    |
| ⚪ **allocs/op** | 8.067k ±1% | 8.063k ±2% | ~ | p=0.782 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/e1f7b69/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/e1f7b69830e195623dab09158d1aa99614614342/bin/pkg_sql_tests benchdiff/e1f7b69/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/e1f7b69/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/2b9d89f/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/2b9d89f2054b4269ae86516951407e84143928df/bin/pkg_sql_tests benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/SQL/3node/oltp_read_write$ --old=2b9d89f --new=e1f7b69 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_read_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.427m ±3% | 3.394m ±2% | ~ | p=0.595 n=15    |
| ⚪ **allocs/op** | 2.083k ±0% | 2.083k ±0% | ~ | p=0.686 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/e1f7b69/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/e1f7b69830e195623dab09158d1aa99614614342/bin/pkg_sql_tests benchdiff/e1f7b69/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/e1f7b69/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/2b9d89f/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/2b9d89f2054b4269ae86516951407e84143928df/bin/pkg_sql_tests benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_read_only$ --old=2b9d89f --new=e1f7b69 ./pkg/sql/tests
```

</details>

</details>
<details><summary><strong>⚪ Sysbench</strong> [KV, 3node, oltp_write_only]</summary>

| Metric                      | Old Commit     | New Commit     | Delta      | Note         |
|-----------------------------|----------------|----------------|------------|--------------|
| ⚪ **sec/op** | 3.694m ±1% | 3.690m ±0% | ~ | p=0.967 n=15    |
| ⚪ **allocs/op** | 4.185k ±0% | 4.184k ±0% | ~ | p=0.067 n=15    |

<details><summary>Reproduce</summary>

**benchdiff binaries**:
```shell
mkdir -p benchdiff/e1f7b69/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/e1f7b69830e195623dab09158d1aa99614614342/bin/pkg_sql_tests benchdiff/e1f7b69/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/e1f7b69/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
mkdir -p benchdiff/2b9d89f/bin/1058449141
gcloud storage cp gs://cockroach-microbench-ci/builds/2b9d89f2054b4269ae86516951407e84143928df/bin/pkg_sql_tests benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
chmod +x benchdiff/2b9d89f/bin/1058449141/cockroachdb_cockroach_pkg_sql_tests
```
**benchdiff command**:
```shell
benchdiff --run=^BenchmarkSysbench/KV/3node/oltp_write_only$ --old=2b9d89f --new=e1f7b69 ./pkg/sql/tests
```

</details>

</details>

<details><summary>Artifacts</summary>

**download**:
```shell
mkdir -p new
gcloud storage cp gs://cockroach-microbench-ci/artifacts/e1f7b69830e195623dab09158d1aa99614614342/20835912438-1/\* new/
mkdir -p old
gcloud storage cp gs://cockroach-microbench-ci/artifacts/2b9d89f2054b4269ae86516951407e84143928df/20835912438-1/\* old/
```

</details>

_built with commit: [e1f7b69830e195623dab09158d1aa99614614342](https://github.com/cockroachdb/cockroach/commit/e1f7b69830e195623dab09158d1aa99614614342)_

<!-- microbench-ci -->
